presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushAggregatesIntoTableScanRuleSet.java

```diff
@@ -0,0 +1,190 @@
+import static io.prestosql.sql.planner.plan.Patterns.aggregation;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+import static java.util.stream.Collectors.toList;
+
+public class PushAggregatesIntoTableScanRuleSet
+{
+    private static final Capture<TableScanNode> TABLE_SCAN = newCapture();
+
+    private static final Pattern<AggregationNode> AGG_WITHOUT_PROJECTION = aggregation().with(source().matching(tableScan().capturedAs(TABLE_SCAN)));
+
+    private TypeManager typeManager;
+    private final Metadata metadata;
+
+    public PushAggregatesIntoTableScanRuleSet(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.typeManager = new InternalTypeManager(metadata);
+    }
+
+    public Set<Rule<AggregationNode>> getRules()
+    {
+        return ImmutableSet.of(new PushAggregationToTableScanWithNoProjection());
+    }
+
+    private class PushAggregationToTableScanWithNoProjection
+            extends BasePushAggregationToTableScan
+    {
+        @Override
+        public Pattern<AggregationNode> getPattern()
+        {
+            return AGG_WITHOUT_PROJECTION;
+        }
+    }
+
+    private abstract class BasePushAggregationToTableScan
+            implements Rule<AggregationNode>
+    {
+        @Override
+        public boolean isEnabled(Session session)
+        {
+            return isEnableAggPushdown(session);
+        }
+
+        @Override
+        public Result apply(AggregationNode node, Captures captures, Context context)
+        {
+            TableScanNode tableScan = captures.get(TABLE_SCAN);
+            Map<String, ColumnHandle> assignments = tableScan.getAssignments()
+                    .entrySet().stream()
+                    .collect(toImmutableMap(entry -> entry.getKey().getName(), Map.Entry::getValue));
+            Map<Symbol, AggregationNode.Aggregation> aggregations = node.getAggregations();
+
+            ImmutableList<AggregateFunction> aggregateFunctions = toAggFunction(aggregations);
+
+            if (aggregateFunctions.isEmpty()) {
+                return Result.empty();
+            }
+
+            GroupingSetDescriptor groupingSets = node.getGroupingSets();
+
+            if (groupingSets.getGroupingSetCount() > 1) {
+                // Let the PushAggregationToTableScanWithGroupingSets rule handle this.
+                return Result.empty();
+            }
+
+            List<ColumnHandle> groupByColumns = groupingSets.getGroupingKeys().stream()
+                    .map(assignments::get)
+                    .collect(toList());
+
+            Optional<AggregatePushdownResult<TableHandle>> connectorTableHandleAggregatePushdownResult = metadata.applyAggregation(context.getSession(),
+                    tableScan.getTable(),
+                    aggregateFunctions,
+                    assignments,
+                    Optional.of(ImmutableList.of(groupByColumns)));
```

Comment:
If this is always present, do we need Optional?

Reply:
I generally just prefer Optional , for now I have just removed all instances of Optional<List> and converted to List.